volumes.cpp

```diff
@@ -1059,249 +1071,256 @@ bool sqIsEllipsoid(const std::vector< double >& args) {
-  
+// Facet number is 0 if not provided
+SurfaceVolume& makeSurface( const SurfaceCard* card, VolumeCache* v, int facet){
   VolumeCache& cache = default_volume_cache;
   if( v != NULL ){
     cache = *v;
   }
 
   
-  if( cache.contains( card ) ){
+  if( cache.contains( card ) && facet == 0){
     return *cache.get( card );
   }
   else{ 
     // SurfaceCard variables:  mnemonic, args, coord_xform  
     SurfaceVolume* surface;
     const std::string& mnemonic = card->getMnemonic();
     const std::vector< double >& args = card->getArgs(); 
-    
-    if( mnemonic == "so"){
-      surface = new SphereSurface( origin, args.at(0) );
-    }
-    else if( mnemonic == "sx"){
-      surface = new SphereSurface( Vector3d( args.at(0), 0, 0 ), args.at(1) );
-    }
-    else if( mnemonic == "sy"){
-      surface = new SphereSurface( Vector3d( 0, args.at(0), 0 ), args.at(1) );
-    }
-    else if( mnemonic == "sz"){
-      surface = new SphereSurface( Vector3d( 0, 0, args.at(0) ), args.at(1) );
-    }
-    else if( mnemonic == "s" || mnemonic == "sph" ){
-      surface = new SphereSurface( Vector3d( args ), args.at(3) );
-    }
-    else if( mnemonic == "sq" && sqIsEllipsoid(args) ){
-      surface = new EllipsoidSurface( Vector3d( args.at(7), args.at(8), args.at(9) ), Vector3d( args )* (-1/args.at(6)) );
-    }
-    else if( mnemonic == "p"){
-      if( args.size() == 4 ){
-        // plane given as ABCD coefficients (Ax+By+Cz-D=0)
-        Vector3d v(args);
-        surface = new PlaneSurface( v, args.at(3)/v.length() );
-      }
-      else if( args.size() == 9 ){
-        // plane is given at three points in space
-        Vector3d v1(args), v2(args,3), v3(args,6);
-
-        // ordering of the points is arbitrary, but make sure v1 is furthest from origin.
-        if( v1.length() < v2.length() ){
-          std::swap( v1, v2 );
-        }
-        if( v1.length() < v3.length() ){
-          std::swap( v1, v3 );
-        }
-
-        // the normal (up to reversal) of the plane 
-        // The normal may need to be reversed to ensure that the origin
-        // has negative sense, as required by MCNP
-        Vector3d normal = v1.add(-v2).cross( v1.add(-v3) ).normalize();
-
-        // If a ray started from the origin and followed the normal vector,
-        // p is the point where it intersects the plane
-        Vector3d p = normal.scale( v1.dot(normal) );
-
-        // cos of the angle between v1 and normal 
-        double angle = normal.dot( v1.normalize() );
-        
-        // invert the normal if the angle is > 90 degrees, which indicates
-        // that reversal is required.
-        if( angle < 0 ){
-          normal   = -normal;
-        }
-
-        //std::cout << normal << " " << p.length() << " : " << angle << std::endl;
-        surface = new PlaneSurface( normal, p.length() );
 
+    // special function for macrobody facets
+    if( facet != 0 ){
+      surface = FacetSurface( mnemonic, args, facet );
+    }
 
+    else{
```

Comment:
while I see that the logic is hierarchical, the implementation of `else if` might allow you to combine this with the next line, thus maintaining the indentation and reducing the apparent size of the diff